x/oracle/keeper/cache/caches.go

```diff
@@ -0,0 +1,227 @@
+				Validator: msg.Validator,
+			})
+		}
+	}
+	index, _ := k.GetIndexRecentMsg(ctx)
+	for i, b := range index.Index {
+		if b >= block-common.MaxNonce {
+			index.Index = index.Index[i:]
+			break
+		}
+		k.RemoveRecentMsg(ctx, b)
+	}
+	k.SetRecentMsg(ctx, recentMsgs)
+	index.Index = append(index.Index, block)
+	k.SetIndexRecentMsg(ctx, index)
+}
+
+func (c *cacheValidator) add(validators map[string]*big.Int) {
+	for operator, newPower := range validators {
+		if power, ok := c.validators[operator]; ok {
+			if newPower.Cmp(zeroBig) == 0 {
+				delete(c.validators, operator)
+				c.update = true
+			} else if power.Cmp(newPower) != 0 {
+				c.validators[operator].Set(newPower)
+				c.update = true
+			}
+		} else {
+			c.update = true
+			np := *newPower
+			c.validators[operator] = &np
+		}
+	}
+}
+
+func (c *cacheValidator) commit(ctx sdk.Context, k common.KeeperOracle) {
+	block := uint64(ctx.BlockHeight())
+	k.SetValidatorUpdateBlock(ctx, types.ValidatorUpdateBlock{Block: block})
+}
+
+func (c *cacheParams) add(p *common.Params) {
+	// params' update is triggered when params is actually updated, so no need to do comparison here, just udpate and mark the flag
+	// TODO: add comparison check, that's something should be done for validation
+	c.params = p
+	c.update = true
+}
+
+func (c *cacheParams) commit(ctx sdk.Context, k common.KeeperOracle) {
+	block := uint64(ctx.BlockHeight())
+	index, _ := k.GetIndexRecentParams(ctx)
+	for i, b := range index.Index {
+		if b >= block-common.MaxNonce {
+			index.Index = index.Index[i:]
+			break
+		}
+		k.RemoveRecentParams(ctx, b)
+	}
+	// remove and append for KVStore
+	k.SetIndexRecentParams(ctx, index)
+	index.Index = append(index.Index, block)
+	k.SetIndexRecentParams(ctx, index)
+}
+
+// memory cache
+// func (c *Cache) AddCache(i any, k common.KeeperOracle) {
+func (c *Cache) AddCache(i any) {
+	switch item := i.(type) {
+	case *ItemM:
+		c.msg.add(item)
+		//	case *params:
+	case ItemP:
+		c.params.add(item)
+	case ItemV:
+		c.validators.add(item)
+	default:
+		panic("no other types are support")
+	}
+}
+
+// func (c *Cache) RemoveCache(i any, k common.KeeperOracle) {
+func (c *Cache) RemoveCache(i any) {
+	if item, isItemM := i.(*ItemM); isItemM {
+		c.msg.remove(item)
+	}
+}
+
+func (c *Cache) GetCache(i any) bool {
+	switch item := i.(type) {
+	case ItemV:
+		if item == nil {
+			return false
+		}
+		for addr, power := range c.validators.validators {
+			item[addr] = power
+		}
+	case ItemP:
+		if item == nil {
+			return false
+		}
+		*item = *(c.params.params)
+	case *[]*ItemM:
+		if item == nil {
+			return false
+		}
+		tmp := make([]*ItemM, 0, len(c.msg))
+		for _, msgs := range c.msg {
+			tmp = append(tmp, msgs...)
+		}
+		*item = tmp
+	default:
+		return false
+	}
+	return true
+}
+
+func (c *Cache) CommitCache(ctx sdk.Context, reset bool, k common.KeeperOracle) {
+	if len(c.msg) > 0 {
+		c.msg.commit(ctx, k)
+		c.msg = make(map[uint64][]*ItemM)
+	}
+
+	if c.validators.update {
```

Comment:
This is not set in `GetCache` so is it true in the `EndBlocker` ? Not sure if I missed something.